### PR TITLE
[ROCM] add device path and use it to setup device

### PR DESCRIPTION
**Pull request by nirvedhmeshram**
- Add device path for rocm
- Setup device using device path
- Fix a bug in context setup
- Change a few hip API calls to their newer counterparts

### CODE REVIEW

File: experimental/rocm/rocm_driver.c [hunk 1]
```diff
@@ -100,11 +100,18 @@ static uint8_t* iree_hal_rocm_populate_device_info(
...
+  hipUUID device_uuid;
   ROCM_IGNORE_ERROR(syms,
                     hipDeviceGetName(device_name, sizeof(device_name), device));
+  ROCM_IGNORE_ERROR(syms, hipDeviceGetUuid(&device_uuid, device));
   memset(out_device_info, 0, sizeof(*out_device_info));
   out_device_info->device_id = (iree_hal_device_id_t)device;
 
+  iree_string_view_t device_path_string =
+      iree_make_string_view(device_uuid.bytes, 16);
+  buffer_ptr += iree_string_view_append_to_buffer(
```
Comment (benvanik) [inline experimental/rocm/rocm_driver.c:130]: these aren't strings and you need to convert them to hex - see what the cuda driver does

File: experimental/rocm/rocm_driver.c [hunk 2]
```diff
@@ -264,18 +271,71 @@ static iree_status_t iree_hal_rocm_driver_create_device_by_id(
...
+                         hipDeviceGetUuid(&device_uuid_fetched, device),
+                         "hipDeviceGetUuid");
+    if (device_uuid_fetched.bytes[0] != '\0' &&
+        memcmp(device_uuid, device_uuid_fetched.bytes, (size_t)(16)) == 0) {
+      break;
+    }
+  }
+
+  if (device_uuid_fetched.bytes[0] == '\0') {
+    return iree_make_status(IREE_STATUS_NOT_FOUND,
```
Comment (benvanik) [inline experimental/rocm/rocm_driver.c:362]: set status here

File: experimental/rocm/rocm_driver.c [hunk 3]
```diff
@@ -264,18 +271,71 @@ static iree_status_t iree_hal_rocm_driver_create_device_by_id(
...
+    ROCM_RETURN_IF_ERROR(&driver->syms,
+                         hipDeviceGetUuid(&device_uuid_fetched, device),
+                         "hipDeviceGetUuid");
+    if (device_uuid_fetched.bytes[0] != '\0' &&
+        memcmp(device_uuid, device_uuid_fetched.bytes, (size_t)(16)) == 0) {
+      break;
+    }
+  }
+
+  if (device_uuid_fetched.bytes[0] == '\0') {
```
Comment (benvanik) [inline experimental/rocm/rocm_driver.c]: only if iree_status_is_ok

File: experimental/rocm/rocm_driver.c [hunk 4]
```diff
@@ -264,18 +271,71 @@ static iree_status_t iree_hal_rocm_driver_create_device_by_id(
...
+    return iree_make_status(IREE_STATUS_NOT_FOUND,
+                            "ROCM device with deviceUUID %s not found",
+                            device_uuid);
+  }
+
+  status = iree_hal_rocm_driver_create_device_by_id(
```
Comment (benvanik) [inline experimental/rocm/rocm_driver.c]: only if iree_status_is_ok

File: experimental/rocm/rocm_driver.c [hunk 5]
```diff
@@ -264,18 +271,71 @@ static iree_status_t iree_hal_rocm_driver_create_device_by_id(
...
+        host_allocator, out_device);
   }
-  return iree_hal_rocm_driver_create_device_by_id(
-      base_driver, IREE_HAL_DEVICE_ID_DEFAULT, param_count, params,
-      host_allocator, out_device);
+
+  // assumption: path => device uuid
+  uint8_t device_uuid[16] = {0};
+  if (device_path.size > 0) {
+    for (int i = 0; i < 16; i++) device_uuid[i] = device_path.data[i];
```
Comment (benvanik) [inline experimental/rocm/rocm_driver.c]: these aren't strings - see what the cuda driver does

**Comment by nirvedhmeshram** [comment]
@benvanik please review again when you get the chance.

**Review by antiagainst — APPROVED** [review]

**Pull request merged**